### ISSUE TITLE
[BUG] Undefined behavior in thread pool's task queue

### ISSUE THREAD

**Issue by dictoon** [OpenImageIO/oiio#2317]
**Describe the bug**
Program crashes ("Access violation executing location...") as instruction pointer jumps to no man's land.

**To Reproduce**

```
OIIO::thread_pool thread_pool;
```

This single line will crash under VS 2015 (19.00.24234.1) with whole-program optimizations and/or other aggressive optimizations enabled.

No crash was observed with other VS versions (e.g. 2017) or without aggressive optimizations.

**Expected behavior**
No crash 😂🍺

**Evidence**
In `src/libutil/thread.cpp` on [lines 55-64](https://github.com/OpenImageIO/oiio/blob/master/src/libutil/thread.cpp#L55-L64):

```
// deletes the retrieved element, do not use for non integral types
bool pop(T& v)
{
    std::unique_lock<Mutex> lock(this->mutex);
    if (this->q.empty())
        return false;
    v = this->q.front();
    this->q.pop();
    return true;
}
```

Even though the comment acknowledges the risk, this code is suspicious, and clearly it is undefined behavior to use the value `v` returned by `pop()`.

That being said, I'm not sure this is the cause of the crash. All I can say for sure at this point is that it crashes with the following callstack:

    -> thread_pool::Impl::set_thread()
       thread_pool::Impl::resize()
       thread_pool::Impl::Impl()

It appears to crash in the newly spawned thread but I don't have more details yet.

**Comment by dictoon** [comment]
I wrote this bug report after many hours of tracing through optimized assembly; with the fresh mind of a new day, I realize that there is no UB in `pop()` as the front value from the queue is simply copied into `v`, and as long as `T` is an integral type then the cost of the copy is pretty much irrelevant.

There is still a crash in `thread_pool::Impl::set_thread()`. I can confirm that the crash occurs when spawning the new thread, i.e. the following line added at the beginning of `thread_pool::Impl::Impl()` is enough:

    new std::thread([]{});

**Comment by lgritz** [comment]
I wonder if there is something funny about MS's implementation of std::function that is causing a problem with the fact that it's being copied.

What happens when you change

    v = this->q.front();

to this alternative:

    v = std::move(this->q.front());

**Comment by dictoon** [comment]
`T` is just a pointer so it should be fine.

I've been able to narrow the problem down further: https://developercommunity.visualstudio.com/content/problem/701555/stdthread-with-lambda-crashes-when-compiled-withou.html

TL;DR: we compile OIIO without Whole Program Optimization (/GL) but appleseed is compiled with (in Ship configuration), and that seems to cause

    new std::thread([]{});

to crash.

**Comment by lgritz** [comment]
Sounds like a MSVS compiler bug? Or are we specifically doing something that results in UB?

**Comment by dictoon** [comment]
It's an MSVC bug I think. That, or /UB + not/UB isn't supported.

**Comment by dictoon** [comment]
MS has [confirmed](https://developercommunity.visualstudio.com/comments/706208/view.html) that this may be an MSVC bug. Using a /GL (with Whole Program Optimization) build of OIIO for our /GL builds of appleseed solved the problem. Good enough for us for now.
